docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,302 @@
+To apply the `default(<default>)` filter one needs to know a default value of
+corresponding role variable (which leads to value duplication that is already set in the role).
+Thus, to eliminate default value duplication and to keep role variables
+definition in the role invocation expression, one can use the `default(omit)`
+filter form:
+
+```yaml
+- include_role:
+    name: <role>
+  vars:
+    r_<role>_<variable>: "{{ inventory_<variable> | default(omit) }}"
+    r_<role>_<variable>: "{{ inventory_<variable> | default(omit) }}"
+    ...
+```
+
+Currently, the `omit` keyword is correctly interpreted by Ansible only
+when used with module arguments (https://github.com/ansible/ansible/issues/13619).
+For that reason the role has to advocate the following pattern to properly
+interpret "omitted" role variables:
+
+```yaml
+- name: "Set r_openshift_node_is_containerized to default if omited"
+  set_fact:
+    r_openshift_node_is_containerized: "{{ __openshift_node_is_containerized }}"
+  when:
+  - __omit_str in r_openshift_node_is_containerized | string
+```
+
+**Explanation**: The `r_openshift_node_is_containerized` is set to `__openshift_node_is_containerized`
+only when the role variable is set to `omit`. Internally the `omit` corresponds to a string
+accepted by `__omit_place_holder__[\da-e]{40}` regex. Thus, if the role variable
+value contains the `__omit_place_holder__` string (stored inside the `__omit_str` variable),
+the role variable is considered to be "undefined" and thus set to its default value.
+This way, the role variables can be set to non-role variables that are optionally defined
+in a compact way.
+
+Since the `__omit_str` can be used on multiple places (e.g. tasks, templates),
+it is defined in `defaults/main.yml` as `__omit_str: "__{{ omit.split('__')[1] }}__"`.
+
+#### We should check all relevant role variables are properly set
+
+A list of tasks checking role variables can be generated based on the role variable annotations.
+Given a list of role variables and their properties (e.g. if a variable is required, a list of accepted values)
+can change over time, it is important to keep the checks in sync with the `default/main.yml`.
+The following checks can be generated:
+
+```yaml
+- name: "Fail if r_openshift_node_deployment_type is not defined"
+  fail:
+    msg: "r_openshift_node_deployment_type must be specified for this role"
+  when:
+  - r_openshift_node_deployment_type is not defined
+
+- name: Fail if invalid r_openshift_node_deployment_type provided
+  fail:
+    msg: "r_openshift_node_deployment_type can only be set to a single value from ['origin', 'openshift-enterprise']"
+  when:
+  - r_openshift_node_deployment_type is defined
+  - r_openshift_node_deployment_type not in ['origin', 'openshift-enterprise']
+```
+
+All the checks can be stored into `tasks/pre_checks.yml` file and included in the main task file.
+
+### Local variables
+
+All variables set inside a role has to be prefixed with `__` string.
```

Comment:
IMHO, if it is not completely obvious that a private variable is set locally before being used, it should have the role name in the prefix to prevent just this kind of collision. In practice, this would probably mean you only use `__foo` within a single task file where the first thing you do is set it. Otherwise `__role_name_foo`.